ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/vm/VmTraceGenerator.java

```diff
@@ -0,0 +1,198 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.vm;
+
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.processor.TransactionTrace;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.Trace;
+import org.hyperledger.besu.ethereum.core.Gas;
+import org.hyperledger.besu.ethereum.debug.TraceFrame;
+import org.hyperledger.besu.ethereum.vm.Code;
+import org.hyperledger.besu.util.bytes.Bytes32;
+import org.hyperledger.besu.util.bytes.BytesValue;
+import org.hyperledger.besu.util.bytes.BytesValues;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class VmTraceGenerator {
+
+  /**
+   * Generate a stream of trace result objects.
+   *
+   * @param transactionTrace the transaction trace to use.
+   * @return a representation of generated traces.
+   */
+  public static Stream<Trace> generateTraceStream(final TransactionTrace transactionTrace) {
+    return Stream.of(generateTrace(transactionTrace));
+  }
+
+  /**
+   * Generate trace representation from the specified transaction trace.
+   *
+   * @param transactionTrace the transaction trace to use.
+   * @return a representation of the trace.
+   */
+  public static Trace generateTrace(final TransactionTrace transactionTrace) {
+    final VmTrace rootVmTrace = new VmTrace();
+    final Deque<VmTrace> parentTraces = new ArrayDeque<>();
+    parentTraces.add(rootVmTrace);
+    if (transactionTrace != null && !transactionTrace.getTraceFrames().isEmpty()) {
+      transactionTrace
+          .getTransaction()
+          .getInit()
+          .map(BytesValue::getHexString)
+          .ifPresent(rootVmTrace::setCode);
+      final AtomicInteger index = new AtomicInteger(0);
+      transactionTrace
+          .getTraceFrames()
+          .forEach(traceFrame -> addFrame(index, transactionTrace, traceFrame, parentTraces));
+    }
+    return rootVmTrace;
+  }
+
+  /**
+   * Add a trace frame to the VmTrace result object.
+   *
+   * @param index index of the current frame in the trace
+   * @param transactionTrace the transaction trace
+   * @param traceFrame the current trace frame
+   * @param parentTraces the queue of parent traces
+   */
+  private static void addFrame(
+      final AtomicInteger index,
+      final TransactionTrace transactionTrace,
+      final TraceFrame traceFrame,
+      final Deque<VmTrace> parentTraces) {
+    if ("STOP".equals(traceFrame.getOpcode())) {
```

Comment:
Why is there a special case for STOP?  Is this an artifact [this EVM logic](https://github.com/hyperledger/besu/blob/e47edb584a6330b862d1e5bc4d7f47cc3d044a04/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/EVM.java#L146-L149)?  If there's an explicit stop in the code executing, does this break?

Reply:
No i don't think it will break. `STOP` opcodes are just ignored by Parity, they don't have any representation in the trace.

Reply:
None of the contracts in the tests have an explicit `STOP` opcode.  Can you add some tests to see what happens?

Reply:
I was wrong. Parity ignores `STOP` only if it is the only opcode. Nice catch !! I have updated test cases to cover this scenario.

Reply:
fixed here: https://github.com/hyperledger/besu/pull/4/commits/710def6c174f1e239bcc79248d1755567b3a9fb4